File: academy/scripts/dojo-agent.ts
import { createHash, createSign, generateKeyPairSync, randomUUID } from 'crypto';
import {
  createAccept,
  createAssess,
  createIntent,
} from '../../itp/src/protocol.ts';
import { IntentSpaceClient } from '../../intent-space/src/client.ts';
import type { ClientTarget, StoredMessage, TlsClientTarget } from '../../intent-space/src/types.ts';
import {
  REGISTRATION_INTENT_CONTENT,
  REGISTRATION_SPACE_ID,
  RITUAL_GREETING_CONTENT,
  TUTORIAL_SPACE_ID,
} from '../src/station-contract.ts';

interface Options {
  target: ClientTarget;
  agentId: string;
  agentName: string;
}

function parseArgs(argv: string[]): Options {
  const opts = new Map<string, string>();
  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    if (!arg.startsWith('--')) continue;
    const key = arg.slice(2);
    const next = argv[i + 1];
    if (!next || next.startsWith('--')) {
      opts.set(key, 'true');
      continue;
    }
    opts.set(key, next);
    i += 1;
  }

  const agentId = opts.get('agent-id') ?? `dojo-agent-${randomUUID().slice(0, 8)}`;
  const agentName = opts.get('agent-name') ?? agentId;

  const socketPath = opts.get('socket');
  if (socketPath) {
    return { target: socketPath, agentId, agentName };
  }

  const host = opts.get('host') ?? '127.0.0.1';
  const port = Number(opts.get('port') ?? '4000');
  if (Number.isNaN(port)) throw new Error('Invalid --port value');

  if (opts.get('tls') === 'true') {
    const target: TlsClientTarget = {
      tls: true,
      host,
      port,
      rejectUnauthorized: opts.get('reject-unauthorized') !== 'false',
    };
    return { target, agentId, agentName };
  }

  return { target: { host, port }, agentId, agentName };
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function waitForMessage(
  client: IntentSpaceClient,
  spaceId: string,
  predicate: (msg: StoredMessage) => boolean,
  description: string,
  timeoutMs = 10_000,
): Promise<StoredMessage> {
  const startedAt = Date.now();
  while (Date.now() - startedAt < timeoutMs) {
    const messages = await client.scan(spaceId);
    const found = messages.find(predicate);
    if (found) return found;
    await sleep(150);
  }
  throw new Error(`Timed out waiting for ${description} in space ${spaceId}`);
}

function logStep(step: string, detail?: string): void {
  console.log(`dojo-agent: ${step}${detail ? ` ${detail}` : ''}`);
}

async function main(): Promise<void> {
  const options = parseArgs(process.argv.slice(2));
  const client = new IntentSpaceClient(options.target);

  const { privateKey, publicKey } = generateKeyPairSync('rsa', { modulusLength: 4096 });
  const publicKeyPem = publicKey.export({ type: 'spki', format: 'pem' }).toString();
  const privateKeyPem = privateKey.export({ type: 'pkcs8', format: 'pem' }).toString();
  const fingerprint = createHash('sha256').update(publicKeyPem).digest('base64');

  try {
    await client.connect();
    logStep('connected', `(agent=${options.agentId})`);

    const registrationIntent = createIntent(options.agentId, REGISTRATION_INTENT_CONTENT);
    registrationIntent.intentId = `registration-${randomUUID()}`;
    registrationIntent.parentId = REGISTRATION_SPACE_ID;
    registrationIntent.payload.agentName = options.agentName;
    registrationIntent.payload.publicKeyPem = publicKeyPem;
    registrationIntent.payload.fingerprint = fingerprint;
    registrationIntent.payload.capabilities = ['scan', 'post', 'enter', 'sign-challenge'];
    registrationIntent.payload.academyVersion = 'phase1';

    client.post(registrationIntent);
    logStep('posted registration intent', registrationIntent.intentId);

    const challenge = await waitForMessage(
      client,
      registrationIntent.intentId,
      (msg) => msg.type === 'INTENT' && typeof msg.payload.challenge === 'string',
      'registration challenge',
    );
    logStep('received challenge');

    const sign = createSign('RSA-SHA256');
    sign.update(String(challenge.payload.challenge));
    sign.end();
    const signatureBase64 = sign.sign(privateKeyPem).toString('base64');

    const signedResponse = createIntent(options.agentId, 'Signed challenge response');
    signedResponse.intentId = `registration-response-${randomUUID()}`;
    signedResponse.parentId = registrationIntent.intentId;
    signedResponse.payload.challenge = challenge.payload.challenge;
    signedResponse.payload.signatureBase64 = signatureBase64;
    client.post(signedResponse);
    logStep('posted signed challenge response');

    await waitForMessage(
      client,
      registrationIntent.intentId,
      (msg) => msg.type === 'INTENT'
        && msg.senderId === 'differ-tutor'
        && msg.payload.tutorialSpaceId === TUTORIAL_SPACE_ID
        && msg.payload.ritualGreeting === RITUAL_GREETING_CONTENT,
      'registration acknowledgement',
    );
    logStep('registration acknowledged');

    const greeting = createIntent(options.agentId, RITUAL_GREETING_CONTENT);
    greeting.intentId = `tutorial-greeting-${randomUUID()}`;
    greeting.parentId = TUTORIAL_SPACE_ID;
    client.post(greeting);
    logStep('posted ritual greeting', greeting.intentId);

    await waitForMessage(
      client,
      greeting.intentId,
      (msg) => msg.type === 'INTENT'
        && msg.senderId === 'differ-tutor'
        && msg.payload.nextStep === 'enter-subspace',
      'tutorial instruction',
    );
    logStep('entered greeting subspace');

    const firstAsk = createIntent(options.agentId, 'I want to try the first tutorial move');
    firstAsk.intentId = `tutorial-first-${randomUUID()}`;
    firstAsk.parentId = greeting.intentId;
    client.post(firstAsk);
    logStep('posted first tutorial intent');

    await waitForMessage(
      client,
      greeting.intentId,
      (msg) => msg.type === 'DECLINE' && msg.intentId === firstAsk.intentId,
      'tutorial decline',
    );
    logStep('received deliberate decline');

    const correctedAsk = createIntent(
      options.agentId,
      'Please guide me through the station ritual with an explicit promise I can accept.',
    );
    correctedAsk.intentId = `tutorial-corrected-${randomUUID()}`;
    correctedAsk.parentId = greeting.intentId;
    client.post(correctedAsk);
    logStep('posted corrected tutorial intent');

    const promise = await waitForMessage(
      client,
      greeting.intentId,
      (msg) => msg.type === 'PROMISE' && typeof msg.promiseId === 'string',
      'tutorial promise',
    );
    logStep('received promise', promise.promiseId);

    const accept = createAccept(options.agentId, promise.promiseId!);
    accept.parentId = greeting.intentId;
    client.post(accept);
    logStep('accepted promise');

    await waitForMessage(
      client,
      greeting.intentId,
      (msg) => msg.type === 'COMPLETE' && msg.promiseId === promise.promiseId,
      'tutorial complete',
    );
    logStep('received complete');

    const assess = createAssess(options.agentId, promise.promiseId!, 'FULFILLED');
    assess.parentId = greeting.intentId;
    client.post(assess);
    logStep('posted assess FULFILLED');

    await waitForMessage(
      client,
      greeting.intentId,
      (msg) => msg.type === 'INTENT'
        && msg.senderId === 'differ-tutor'
        && msg.payload.content === 'Tutorial complete. You can now proceed beyond the ritual.',
      'final tutorial acknowledgement',
    );

    logStep('happy path complete');
  } finally {
    client.disconnect();
  }
}

main().catch((error: unknown) => {
  const message = error instanceof Error ? error.message : String(error);
  console.error(`dojo-agent: failed ${message}`);
  process.exit(1);
});

File: academy/scripts/dojo-agent.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import sys
from pathlib import Path
from typing import Any


SCRIPT_DIR = Path(__file__).resolve().parent
SKILL_SDK_DIR = SCRIPT_DIR.parent / "skill-pack" / "sdk"
sys.path.insert(0, str(SKILL_SDK_DIR))

from promise_runtime import PromiseRuntimeSession  # noqa: E402


TUTORIAL_SPACE_ID = "tutorial"
RITUAL_GREETING_CONTENT = "academy tutorial greeting"


def parse_args(argv: list[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--host", default="127.0.0.1")
    parser.add_argument("--port", type=int, default=4000)
    parser.add_argument("--academy-url", default="http://127.0.0.1:8080")
    parser.add_argument("--agent-id", default=None)
    parser.add_argument("--agent-name", default=None)
    parser.add_argument("--workspace", default=None)
    return parser.parse_args(argv)


def log(step: str, detail: str | None = None) -> None:
    print(f"dojo-agent: {step}{f' {detail}' if detail else ''}", flush=True)


def save_if_present(session: PromiseRuntimeSession, filename: str, payload: Any) -> None:
    if isinstance(payload, dict):
        session.save_json_artifact(filename, payload)


def main(argv: list[str]) -> int:
    args = parse_args(argv)
    workspace = Path(args.workspace or SCRIPT_DIR.parent / "tmp" / "dojo-agent")
    agent_name = args.agent_name or args.agent_id or "dojo-agent"
    session = PromiseRuntimeSession(
        endpoint=f"tcp://{args.host}:{args.port}",
        workspace=workspace,
        agent_name=agent_name,
        agent_id=args.agent_id,
    )
    identity = session.identity()
    try:
        signup = session.signup(args.academy_url, handle=session.agent_id)
        endpoint = str(session.endpoint)
        session.connect()
        session.record_step("dojo.start", {"workspace": str(workspace), "endpoint": endpoint, "academyUrl": args.academy_url})
        session.save_json_artifact("session-start.json", session.snapshot())
        log("connected", f"(agent={session.agent_id})")

        root_scan = session.scan("root")
        save_if_present(session, "root-scan.json", root_scan)

        greeting = session.post(
            session.intent(RITUAL_GREETING_CONTENT, parent_id=TUTORIAL_SPACE_ID),
            step="dojo.tutorial.post_greeting",
            artifact_filename="tutorial-greeting.json",
        )
        greeting_space = greeting["intentId"]
        log("posted ritual greeting", greeting_space)

        tutorial_scan = session.wait_for_intent(
            greeting_space,
            sender_id="differ-tutor",
            payload_predicate=lambda payload: payload.get("nextStep") == "enter-subspace",
            wait_seconds=20.0,
        )
        save_if_present(session, "tutorial-scan.json", tutorial_scan)
        session.record_step("dojo.tutorial.greeting_acknowledged", {"intentId": tutorial_scan.get("intentId")})
        log("entered greeting subspace")

        first_ask = session.post(
            session.intent("I want to try the first tutorial move", parent_id=greeting_space),
            step="dojo.tutorial.post_first_attempt",
        )
        log("posted first tutorial intent")

        decline = session.wait_for_decline(
            greeting_space,
            intent_id=first_ask["intentId"],
            wait_seconds=20.0,
        )
        save_if_present(session, "tutorial-decline.json", decline)
        session.record_step("dojo.tutorial.decline_received", {"intentId": decline.get("intentId")})
        log("received deliberate decline")

        corrected = session.post(
            session.intent(
                "Please guide me through the station ritual with an explicit promise I can accept.",
                parent_id=greeting_space,
            ),
            step="dojo.tutorial.post_corrected_intent",
        )
        log("posted corrected tutorial intent")

        promise = session.wait_for_promise(
            greeting_space,
            payload_predicate=lambda payload: isinstance(payload, dict),
            wait_seconds=20.0,
        )
        save_if_present(session, "tutorial-promise.json", promise)
        promise_id = promise["promiseId"]
        session.record_step("dojo.tutorial.promise_received", {"promiseId": promise_id})
        log("received promise", promise_id)

        accept = session.post(
            session.accept(promise_id=promise_id, parent_id=greeting_space),
            step="dojo.tutorial.post_accept",
            artifact_filename="tutorial-accept.json",
        )
        log("accepted promise")

        complete = session.wait_for_complete(
            greeting_space,
            promise_id=promise_id,
            wait_seconds=20.0,
        )
        save_if_present(session, "tutorial-complete.json", complete)
        session.record_step("dojo.tutorial.complete_received", {"promiseId": promise_id})
        log("received complete")

        assess = session.post(
            session.assess(
                promise_id=promise_id,
                parent_id=greeting_space,
                assessment="FULFILLED",
            ),
            step="dojo.tutorial.post_assess",
            artifact_filename="tutorial-assess.json",
        )
        log("posted assess FULFILLED")

        final_ack = session.wait_for_intent(
            greeting_space,
            sender_id="differ-tutor",
            payload_predicate=lambda payload: payload.get("content")
            == "Tutorial complete. You can now proceed beyond the ritual.",
            wait_seconds=20.0,
        )
        save_if_present(session, "tutorial-final-ack.json", final_ack)
        session.record_step("dojo.tutorial.final_ack_received", {"intentId": final_ack.get("intentId")})
        reward = final_ack.get("payload", {}).get("dojoReward") if isinstance(final_ack.get("payload"), dict) else None
        certificate = final_ack.get("payload", {}).get("dojoCertificate") if isinstance(final_ack.get("payload"), dict) else None
        if isinstance(reward, dict):
            session.save_json_artifact("dojo-token.txt.json", reward)
        if isinstance(certificate, dict):
            session.save_json_artifact("dojo-certificate.json", certificate)
        session.record_step("dojo.complete", {"reward": isinstance(reward, dict), "certificate": isinstance(certificate, dict)})
        session.save_json_artifact("session-finish.json", session.snapshot())
        log("happy path complete")
        return 0
    finally:
        session.close()


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
